src/bn254/fq.rs

```diff
@@ -34,6 +38,273 @@ impl Fp254Impl for Fq {
+
+                    let lc_signs = <Fq as [<Fp254 $NAME>]>::LCS;
+
+                    type U = <Fq as [<Fp254 $NAME>]>::U;
+                    type T = <Fq as [<Fp254 $NAME>]>::T;
+
+                    // N_BITS for the extended number used during intermediate computation
+                    const MAIN_LOOP_END: u32 = {
+                        let n_bits_mod_width = ((N_BITS + MOD_WIDTH - 1) / MOD_WIDTH) * MOD_WIDTH;
+                        let n_bits_var_width = ((N_BITS + VAR_WIDTH - 1) / VAR_WIDTH) * VAR_WIDTH;
+                        let mut u = n_bits_mod_width;
+                        if n_bits_var_width > u {
+                            u = n_bits_var_width;
+                        }
+                        while !(u % MOD_WIDTH == 0 && u % VAR_WIDTH == 0) {
+                            u += 1;
+                        }
+                        u
+                    };
+
+                    // Pre-computed lookup table allows us to skip initial few doublings
+                    const MAIN_LOOP_START: u32 = {
+                        if MOD_WIDTH < VAR_WIDTH {
+                            MOD_WIDTH
+                        } else {
+                            VAR_WIDTH
+                        }
+                    };
+
+                    const N_WINDOW: u32 = MAIN_LOOP_END / VAR_WIDTH;
+
+                    // Pre-computed lookup table's size
+                    fn size_table(window: u32) -> u32 { (1 << window) - 1 }
+
+                    // Initialize the lookup table
+                    fn init_table(window: u32) -> Script {
+                        assert!(
+                            1 <= window && window <= 6,
+                            "expected 1<=window<=6; got window={}",
+                            window
+                        );
+                        script! {
+                            for i in 2..=window {
+                                for j in 1 << (i - 1)..1 << i {
+                                    if j % 2 == 0 {
+                                        { T::copy(j/2 - 1) }
+                                        { T::double_allow_overflow() }
+                                    } else {
+                                        { T::copy(0) }
+                                        { T::add_ref(j - 1) }
+                                    }
+                                }
+                            }
+                        }
+                    }
+
+                    // Drop the lookup table
+                    fn drop_table(window: u32) -> Script {
+                        script! {
+                            for _ in 1..1<<window {
+                                { T::drop() }
+                            }
+                        }
+                    }
+
+                    // Get modulus window at given index
+                    fn mod_window(index: u32) -> u32 {
+                        let shift_by = MOD_WIDTH * (N_WINDOW - index - 1);
```

Comment:
I think instead of `const N_WINDOW: u32 = MAIN_LOOP_END / VAR_WIDTH;`, another const `MAIN_LOOP_END / MOD_WIDTH` should be used here for `N_WINDOW`. It does not make a difference when MOD_WIDTH and VAR_WIDTH are the same (I know the best option is when they are equal at 4) but (4, 3) and (3, 4) are not working currently.

Reply:
Yeah, moving the N_WINDOW into respective functions makes sense. For mod_window, it should be MAIN_LOOP_END / MOD_WIDTH, and for var_window_script, it should be MAIN_LOOP_END / VAR_WIDTH.